### ISSUE TITLE
use beam search during training

### ISSUE THREAD

**Issue by zhu-y11** [OpenNMT/OpenNMT-py#1626]
Hi, 

I am trying to generate **multiple candidate sentences** (beam search) during training, and use some sequence level loss to backprop through it. 

However I find it hard for me to integrate beam search code in *translate* to the trainer. Is there a fast way to do that or it's probably better to implement my own beam search?

Thanks a lot! :)

**Comment by chrishokamp** [comment]
hi, a typical hack for this is to just generate outputs `Y_hat` from the model (i.e. sample or translate with beam search), then pass a batch of `(X, Y_hat, Y)` to update parameters. So at each step of training you do:
```
Y_hat = sample(model, X)
update_parameters(model, X, Y_hat, Y)
```
in the `train` step, compute the sequence level loss on `(Y_hat, Y)`, or maybe just on `Y_hat` :wink:

**Comment by zhu-y11** [comment]
> hi, a typical hack for this is to just generate outputs `Y_hat` from the model (i.e. sample or translate with beam search), then pass a batch of `(X, Y_hat, Y)` to update parameters. So at each step of training you do:
> 
> ```
> Y_hat = sample(model, X)
> update_parameters(model, X, Y_hat, Y)
> ```

That seems pretty neat! So in this case I need to first save the checkpoint of the trainer and then load it with a translator, so I will have 2 models in total? I am not sure if you are referring to this or some other way?

**Comment by chrishokamp** [comment]
Create a translator and trainer that use the same model

**Comment by zhu-y11** [comment]
I also found the beam search return empty list for some sentences when the batch_size > 1, in my case, beam_size =9. However, it doesn't happen when the batch_size = 1. Might be related to [#1345](https://github.com/OpenNMT/OpenNMT-py/issues/1345)?
